### PR TITLE
${AssemblyVersion}: add type (File, Assembly, Informational) option

### LINKED ISSUE #2480: Add version types to AssemblyVersion layout renderer
**Type**: Feature request

**NLog version**: 4.5.0

**Platform**: .Net 3.5 / .Net 4.0 / .Net 4.5 / Mono 4 / Silverlight 4 / Silverlight 5 / Xamarin Android / Xamarin iOs / .NET Core 

**Current NLog config**: N/A

I would like to extend the AssemblyVersion layout renderer to be able to specify different version types, e.g. from [here](https://github.com/dotnet/sdk/blob/1f1485fc216cd87b0289737eb7fe7de417335022/src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.GenerateAssemblyInfo.targets#L56-L85):
- AssemblyVersion
- FileVersion
- InformationalVersion

So the current renderer syntax of `${assembly-version:name=string}` could become `${assembly-version:name=string:type=string}` where the value for `type` is one of:
- assembly
- file
- informational

I'm happy to provide this in a pull request with unit tests, etc... if agreed 😃

### PR DESCRIPTION
**Pull request by alexangas**
Fixes #2480.

Requires documentation updates at https://github.com/nlog/nlog/wiki/AssemblyVersion-Layout-Renderer and https://github.com/NLog/NLog/wiki/Platform-Support (see code for details).

### CODE REVIEW

**Comment by 304NotModified** [comment]
The build server says:

```
'AssemblyVersionLayoutRenderer.GetVersion()': not all code paths return a value [C:\projects\nlog\src\NLog\NLog.csproj]

```

Could you please check that?

File: src/NLog/LayoutRenderers/AssemblyVersionLayoutRenderer.cs [hunk 1]
```diff
@@ -31,60 +31,144 @@
...
         }
 
-        string GetAssemblyApplicationVersion(string assemblyName)
+#if SILVERLIGHT
+
+        private string GetVersion()
         {
-            if (!string.IsNullOrEmpty(assemblyName))
+            var assemblyName = GetAssemblyName();
+            assemblyName.Version.ToString();
```
Comment (304NotModified) [inline src/NLog/LayoutRenderers/AssemblyVersionLayoutRenderer.cs]: missing `return` ;)

File: tests/NLog.UnitTests/LayoutRenderers/AssemblyVersionTests.cs [hunk 2]
```diff
@@ -52,5 +53,93 @@ public void AssemblyNameVersionTest()
...
+            var testLoggerType = compiledAssembly.GetType("LogTester.LoggerTest");
+            var logMethod = testLoggerType.GetMethod("TestLog");
+            var testLoggerInstance = Activator.CreateInstance(testLoggerType);
+            logMethod.Invoke(testLoggerInstance, new object[] { logger, compiledAssembly });
+            AssertDebugLastMessage("debug", expected);
+        }
+
+        private static Assembly GenerateTestAssembly()
+        {
+            const string code = @"
```
Comment (304NotModified) [inline tests/NLog.UnitTests/LayoutRenderers/AssemblyVersionTests.cs]: nice!

**Comment by 304NotModified** [comment]
Unfortunately, it fails on Travis

.>  NLog.UnitTests.LayoutRenderers.AssemblyVersionTests.AssemblyVersionTypeTest(type: Assembly, expected: "1.1.1.1") [FAIL]
      System.InvalidOperationException : Unable to find field m_entryAssembly

Any idea why? See https://travis-ci.org/NLog/NLog/builds/324836802

**Comment by alexangas** [comment]
The issue appears to be the way that xunit handles app domains, described in Microsoft/vstest#649. This means that calls to `Assembly.GetEntryAssembly()` may return null (as can happen in existing `EntryAssemblyVersionTest`) or attempts to force set it may fail (as can happen in new `AssemblyVersionTypeTest`).

There may be a way to get around this by changing the xunit config value `xunit.appDomain`, but it looks like this may affect all tests and cause adverse effects.

So in these most recent pushes I have taken the same approach as the existing `EntryAssemblyVersionTest` when this wasn't working where it effectively skips the test if there are problems. It also attempts to log that.

**Comment by 304NotModified** [comment]
> So in these most recent pushes I have taken the same approach as the existing EntryAssemblyVersionTest when this wasn't working where it effectively skips the test if there are problems. It also attempts to log that

seems to work great!

Thanks, looks good! 

Could you please update the wiki and add "Introduced in NLog 4.5"? Thanks in advance!

**Comment by alexangas** [comment]
Really appreciate all your help. I've updated https://github.com/nlog/nlog/wiki/AssemblyVersion-Layout-Renderer and https://github.com/NLog/NLog/wiki/Platform-Support.

**Comment by 304NotModified** [comment]
thanks! also for adding examples! 👍 👍

**Pull request merged**